integration-tests/src/tests/client/features/multinode_stateless_validators.rs

```diff
@@ -0,0 +1,674 @@
+
+    // Give the test a chance to finish off remaining events in the event loop, which can
+    // be important for properly shutting down the nodes.
+    test.shutdown_and_drain_remaining_events(Duration::seconds(20));
+}
+
+/// Returns the CurrentEpochValidatorInfo for each validator account for the given epoch id.
+fn get_current_validators(
+    client: &Client,
+    epoch_id: EpochId,
+) -> HashMap<AccountId, CurrentEpochValidatorInfo> {
+    client
+        .epoch_manager
+        .get_validator_info(ValidatorInfoIdentifier::EpochId(epoch_id))
+        .unwrap()
+        .current_validators
+        .iter()
+        .map(|v| (v.account_id.clone(), v.clone()))
+        .collect()
+}
+
+/// Asserts the following:
+/// 1. Block and chunk producers produce block and chunks and also validate chunk witnesses.
+/// 2. Chunk validators only validate chunk witnesses.
+/// 3. Stake of both the block/chunk producers and chunk validators increase (due to rewards).
+/// TODO: Assert on the specific reward amount, currently it only checks that some amount is rewarded.
+fn assert_validator_info(
+    client: &Client,
+    epoch_id: EpochId,
+    initial_epoch_id: EpochId,
+    accounts: &Vec<AccountId>,
+) {
+    let validator_to_info = get_current_validators(client, epoch_id);
+    let initial_validator_to_info = get_current_validators(client, initial_epoch_id);
+
+    // Check that block/chunk producers generate blocks/chunks and also endorse chunk witnesses.
+    for idx in 0..NUM_BLOCK_AND_CHUNK_PRODUCERS {
+        let account = &accounts[idx];
+        let validator_info = validator_to_info.get(account).unwrap();
+        assert!(validator_info.num_produced_blocks > 0);
+        assert!(validator_info.num_expected_blocks >= validator_info.num_produced_blocks);
+
+        assert!(validator_info.num_produced_chunks > 0);
+        assert!(validator_info.num_expected_chunks >= validator_info.num_produced_chunks);
+
+        assert!(validator_info.num_produced_endorsements > 0);
+        assert!(
+            validator_info.num_expected_endorsements >= validator_info.num_produced_endorsements
+        );
+
+        let initial_validator_info = initial_validator_to_info.get(account).unwrap();
+        assert!(initial_validator_info.stake < validator_info.stake);
+    }
+    // Check chunk validators only endorse chunk witnesses.
+    for idx in NUM_BLOCK_AND_CHUNK_PRODUCERS..NUM_VALIDATORS {
+        let account = &accounts[idx];
+        let validator_info = validator_to_info.get(account).unwrap();
+        assert_eq!(validator_info.num_expected_blocks, 0);
+        assert_eq!(validator_info.num_expected_chunks, 0);
+        assert_eq!(validator_info.num_produced_blocks, 0);
+        assert_eq!(validator_info.num_produced_chunks, 0);
+
+        assert!(validator_info.num_produced_endorsements > 0);
+        assert!(
+            validator_info.num_expected_endorsements >= validator_info.num_produced_endorsements
+        );
```

Comment:
We could also assert that `num_expected_endorsements <= epoch_length * num_shards` as a sanity check. Or something lower, if it is possible.